### ISSUE TITLE
Upgrade to wgpu 22

### ISSUE THREAD

**Issue by hannobraun** [hannobraun/fornjot#2424]
This upgrade requires manual intervention. Here's the build failure from the automatic upgrade attempt in https://github.com/hannobraun/fornjot/pull/2422:
https://github.com/hannobraun/fornjot/actions/runs/10040783745/job/27747509273?pr=2422

Labeling https://github.com/hannobraun/fornjot/labels/good%20first%20issue, as this doesn't require a deep knowledge of Fornjot, those build errors don't look too intimidating, and the recent wgpu upgrades haven't been too involved.

**Comment by felix91gr** [comment]
Lemme try it. If I don't ping back in 24 hours, feel free to assign it to someone else c:

**Comment by felix91gr** [comment]
Okay, I have a candidate fix. Let me test it and then make a PR

**Comment by felix91gr** [comment]
I might need some advice. Here is what I have for now.

### Summary

I have fixes for the 3 build errors that appeared in CI. The three of them, I think, can be dealt with without much fuss.

Further testing does reveal another build error, which I'll describe in the second section.

### Fixes for the 3 errors seen in CI

These are the errors, reproduced on my end:

![image](https://github.com/user-attachments/assets/1749e01f-3e12-4602-8912-65fec950d71f)

The first one is due to the new `MemoryHints` field that is fed to `DeviceDescriptor`. I am currently asking for a `MemoryUsage` profile, on the grounds that it should help with compatibility (I think, I might be wrong on this).

This error is fixed by just adding a memory usage profile.

The second and third errors are due to a new field, `cache: Option<PipelineCache>` that is now a part of `RenderPipelineDescriptor`. I have set it to `None` for now, and can be changed later if more performance is needed.

Those fixes make `cargo build`, `test` and `clippy` all pass without a problem. 

### The new issue I'm looking at (advice is much appreciated)

I ran `just ci` locally, and this is the error I get:

![image](https://github.com/user-attachments/assets/80621200-5bd8-44cc-bd93-0157e44af059)

The `RUSTFLAGS` that were used are the following:

`Running RUSTFLAGS="-D warnings" "cargo" "build" "--all-features" "--target" "wasm32-unknown-unknown" "-p" "fj-viewer"`

I'm not sure what to do about this error.

Reading the release notes, it looks like `enumerate_adapters` is now native-only:

![image](https://github.com/user-attachments/assets/87be7789-0cb5-48af-b1d1-8556c4bb427f)

I'm guessing that means we can't use it on `wasm32-unknown-unknown` anymore, and hence a different approach to obtaining the adapters is needed. `request_adapters`, perhaps? I'll take a look at it tomorrow, and see if something clicks about it. I'm pretty sure something will.

**Comment by hannobraun** [comment]
Thank you for looking into this, @felix91gr!

My reading matches yours: We need to avoid `enumerate_adapters` with `wasm32-unknown-unknown`, and need `request_adapter` instead. This requires a surface, as part of [`RequestAdapterOptions`](https://docs.rs/wgpu/22.0.0/wgpu/type.RequestAdapterOptions.html), but that's not a problem, since we already [have one where we need it](https://github.com/hannobraun/fornjot/blob/41331a8185d2f03cea43f4b28ef8b24c519eef51/crates/fj-viewer/src/graphics/renderer.rs#L47).

Looking over the code, I think this shouldn't be too hard to adapt. Let's look at what we're doing:

1. [Use `enumerate_adapters` to log all adapters before trying to do anything.](https://github.com/hannobraun/fornjot/blob/41331a8185d2f03cea43f4b28ef8b24c519eef51/crates/fj-viewer/src/graphics/renderer.rs#L49-L51) Should be fine to just slap a `#[cfg(not(target = "wasm32-unknown-unknown"))]` on there. (Not sure about the specific syntax, but something along those lines should work.
2. [Call `Device::from_preferred_adapter`](https://github.com/hannobraun/fornjot/blob/41331a8185d2f03cea43f4b28ef8b24c519eef51/crates/fj-viewer/src/graphics/renderer.rs#L53), which uses `request_adapter` internally. This is the main code path and should require no change.
3. In case of failure, [call `Device::try_from_all_adapters` as a fallback](https://github.com/hannobraun/fornjot/blob/41331a8185d2f03cea43f4b28ef8b24c519eef51/crates/fj-viewer/src/graphics/renderer.rs#L59). Since this is just fallback behavior, it should be fine to just slap another `#[cfg(not(...))]` on there.

So to summarize, the main code path is already alright, and just debug output and fallback in case of errors use the problematic `enumerate_adapters`. Since, according the the changelog (as highlighted in the screenshot you posted), `enumerate_adapters` returned an empty `Vec` all along on WASM, just configuring them out as I suggested should not change any behavior.

**Comment by hannobraun** [comment]
Addressed in https://github.com/hannobraun/fornjot/pull/2427.
